// vuetify/es5/components/Vuetify/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _application = require('./mixins/application');

var _application2 = _interopRequireDefault(_application);

var _theme = require('./mixins/theme');

var _theme2 = _interopRequireDefault(_theme);

var _options = require('./mixins/options');

var _options2 = _interopRequireDefault(_options);

var _goTo = require('./util/goTo');

var _goTo2 = _interopRequireDefault(_goTo);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Vuetify = {
  install: function install(Vue) {
    var opts = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    if (this.installed) return;

    this.installed = true;

    var $vuetify = {};
    Vue.util.defineReactive($vuetify, 'inspire', {
      breakpoint: {},
      application: _application2.default,
      dark: false,
      theme: (0, _theme2.default)(opts.theme),
      options: (0, _options2.default)(opts.options),
      goTo: _goTo2.default
    });

    Vue.prototype.$vuetify = $vuetify.inspire;

    if (opts.transitions) {
      Object.values(opts.transitions).forEach(function (transition) {
        if (transition.name !== undefined && transition.name.startsWith('v-')) {
          Vue.component(transition.name, transition);
        }
      });
    }

    if (opts.directives) {
      Object.values(opts.directives).forEach(function (directive) {
        Vue.directive(directive.name, directive);
      });
    }

    if (opts.components) {
      Object.values(opts.components).forEach(function (component) {
        Vue.use(component);
      });
    }
  }
};

exports.default = Vuetify;
